### PR TITLE
1 setup online store

### PR DESCRIPTION
**Pull request by AnaLanda**

### CODE REVIEW

File: src/main/java/dao/impl/ProductDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            originalProduct = originalProductOptional.get();
+            int index = Storage.products.indexOf(originalProduct);
+            Storage.products.set(index, product);
+        }
+        return product;
+    }
+
+    @Override
+    public boolean deleteById(Long productId) {
+        Storage.products.removeIf(x -> x.getId().equals(productId));
```
Comment (aleksandr-hrankin) [inline src/main/java/dao/impl/ProductDaoImpl.java]: метод removeIf возвращает boolean)

File: src/main/java/dao/impl/ProductDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    public Product update(Product product) {
+        Product originalProduct = null;
+        Optional<Product> originalProductOptional = Storage.products.stream()
+                .filter(x -> x.getId().equals(product.getId()))
+                .findFirst();
+        if (originalProductOptional.isPresent()) {
+            originalProduct = originalProductOptional.get();
+            int index = Storage.products.indexOf(originalProduct);
+            Storage.products.set(index, product);
+        }
```
Comment (aleksandr-hrankin) [inline src/main/java/dao/impl/ProductDaoImpl.java]: Можно обойтись без Optional, сразу возвращать product

File: src/main/java/dao/impl/ProductDaoImpl.java (hunk 1 above)
Comment (AnaLanda) [reply src/main/java/dao/impl/ProductDaoImpl.java]: Спасибо, Саш! Недосмотрела)

File: pom.xml [hunk 3]
```diff
@@ -8,5 +8,60 @@
...
+                    <configuration>
+                        <source>${jdk.version}</source>
+                        <target>${jdk.version}</target>
+                        <encoding>${project.build.sourceEncoding}</encoding>
+                    </configuration>
+                </plugin>
+            </plugins>
+        </pluginManagement>
+    </build>
 </project>
```
Comment (dmytro-hryhoruk) [inline pom.xml]: add empty line to the end of file

File: src/main/java/com/internet/shop/db/Storage.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+import java.util.List;
+
+public class Storage {
+    public static final List<Product> products = new ArrayList<>();
+    private static Long productId = 0L;
+
+    public static void addProduct(Product product) {
+        productId++;
+        product.setId(productId);
+        products.add(product);
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/db/Storage.java]: let's first set id and then increment it so the first product Id will be zero

File: src/main/java/com/internet/shop/model/Order.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package com.internet.shop.model;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Order {
+    Long id;
+    List<Product> products = new ArrayList<>();
+    Long userId;
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/model/Order.java]: access modifiers

File: src/main/java/com/internet/shop/model/Order.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package com.internet.shop.model;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Order {
+    Long id;
+    List<Product> products = new ArrayList<>();
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/model/Order.java]: let's move this list initialization to constructor

File: src/main/java/com/internet/shop/model/Product.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
+package com.internet.shop.model;
+
+public class Product {
+    Long id;
+    String name;
+    double price;
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/model/Product.java]: access modifier

File: src/main/java/com/internet/shop/model/ShoppingCart.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package com.internet.shop.model;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ShoppingCart {
+    Long id;
+    List<Product> products = new ArrayList<>();
+    Long userId;
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/model/ShoppingCart.java]: same

File: src/main/java/com/internet/shop/model/User.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
+package com.internet.shop.model;
+
+public class User {
+    Long id;
+    String name;
+    String login;
+    String password;
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/model/User.java]: and here

Comment (AnaLanda) [reply src/main/java/com/internet/shop/model/User.java]: :woman_facepalming:

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+                .findFirst();
+    }
+
+    @Override
+    public Product update(Product product) {
+        return Storage.products.stream()
+                .filter(originalProduct -> originalProduct.getId().equals(product.getId()))
+                .map(originalProduct -> product)
+                .findFirst()
+                .orElse(null);
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: let's check this implementation. print all Storage list before updating this product and after

File: pom.xml [hunk 11]
```diff
@@ -8,5 +8,60 @@
...
-</project>
\ No newline at end of file
+    <dependencies>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.12</version>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
```
Comment (KseniiaMakarova) [inline pom.xml]: let's remove `<dependencies>` block for now - we don't use dependencies yet

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java (hunk 10 above)
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: this doesn't cause changes in our storage
the elements are changed only in this stream

let's create IntStream of indices, than filter out only those indices on which there are elements with the same id
then for each of the remaining indices, replace the element in storage with our `product`

also, return `product` in each case

File: src/main/java/com/internet/shop/model/Order.java [hunk 12]
```diff
@@ -0,0 +1,15 @@
+package com.internet.shop.model;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Order {
+    private Long id;
+    private List<Product> products;
+    private Long userId;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/Order.java:9]: getters and setters would be nice :)

File: src/main/java/com/internet/shop/model/ShoppingCart.java [hunk 13]
```diff
@@ -0,0 +1,15 @@
+package com.internet.shop.model;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ShoppingCart {
+    private Long id;
+    private List<Product> products;
+    private Long userId;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/ShoppingCart.java:9]: same here

File: src/main/java/com/internet/shop/model/User.java [hunk 14]
```diff
@@ -0,0 +1,14 @@
+package com.internet.shop.model;
+
+public class User {
+    private Long id;
+    private String name;
+    private String login;
+    private String password;
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/model/User.java:7]: same here

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
also, in this PR let's fix the last line of your `gitignore`

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    @Override
+    public Product update(Product product) {
+        List<Product> products = getAllProducts();
+        IntStream.range(0, products.size())
+                .filter(originalProduct
+                        -> products.get(originalProduct).getId().equals(product.getId()))
+                .findFirst()
+                .ifPresent(originalProduct -> products.set(originalProduct, product));
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]:
```suggestion
        IntStream.range(0, products.size())
                .filter(index
                        -> products.get(index).getId().equals(product.getId()))
                .forEach(index -> products.set(index, product));
```

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
please fix last line in your `gitignore`

it is currently `target/*git`, should be `target/*`

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java (hunk 15 above)
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]:
```suggestion
                .filter(i -> products.get(i).getId().equals(product.getId()))
                .forEach(i -> products.set(i, product));
```

Comment (AnaLanda) [reply src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: Thanks))

Comment (AnaLanda) [reply src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: Thanks))

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**